checker/src/org/checkerframework/checker/upperbound/UpperBoundTransfer.java

```diff
@@ -0,0 +1,440 @@
+package org.checkerframework.checker.upperbound;
+
+import java.util.List;
+import java.util.Set;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.type.TypeKind;
+import org.checkerframework.checker.upperbound.qual.LTEqLengthOf;
+import org.checkerframework.checker.upperbound.qual.LTLengthOf;
+import org.checkerframework.checker.upperbound.qual.LTOMLengthOf;
+import org.checkerframework.dataflow.analysis.ConditionalTransferResult;
+import org.checkerframework.dataflow.analysis.FlowExpressions;
+import org.checkerframework.dataflow.analysis.FlowExpressions.FieldAccess;
+import org.checkerframework.dataflow.analysis.FlowExpressions.Receiver;
+import org.checkerframework.dataflow.analysis.RegularTransferResult;
+import org.checkerframework.dataflow.analysis.TransferInput;
+import org.checkerframework.dataflow.analysis.TransferResult;
+import org.checkerframework.dataflow.cfg.node.ArrayCreationNode;
+import org.checkerframework.dataflow.cfg.node.AssignmentNode;
+import org.checkerframework.dataflow.cfg.node.EqualToNode;
+import org.checkerframework.dataflow.cfg.node.FieldAccessNode;
+import org.checkerframework.dataflow.cfg.node.GreaterThanNode;
+import org.checkerframework.dataflow.cfg.node.GreaterThanOrEqualNode;
+import org.checkerframework.dataflow.cfg.node.LessThanNode;
+import org.checkerframework.dataflow.cfg.node.LessThanOrEqualNode;
+import org.checkerframework.dataflow.cfg.node.Node;
+import org.checkerframework.dataflow.cfg.node.NotEqualNode;
+import org.checkerframework.framework.flow.CFAbstractTransfer;
+import org.checkerframework.framework.flow.CFValue;
+import org.checkerframework.framework.type.QualifierHierarchy;
+import org.checkerframework.javacutil.AnnotationUtils;
+
+public class UpperBoundTransfer
+        extends CFAbstractTransfer<CFValue, UpperBoundStore, UpperBoundTransfer> {
+
+    private UpperBoundAnnotatedTypeFactory atypeFactory;
+
+    private QualifierHierarchy qualifierHierarchy;
+
+    public UpperBoundTransfer(UpperBoundAnalysis analysis) {
+        super(analysis);
+        atypeFactory = (UpperBoundAnnotatedTypeFactory) analysis.getTypeFactory();
+        qualifierHierarchy = atypeFactory.getQualifierHierarchy();
+    }
+
+    // Refine the type of expressions used as an array dimension to be
+    // less than length of the array to which the new array is
+    // assigned.
+    @Override
+    public TransferResult<CFValue, UpperBoundStore> visitAssignment(
+            AssignmentNode node, TransferInput<CFValue, UpperBoundStore> in) {
+        AnnotationMirror UNKNOWN = atypeFactory.UNKNOWN;
+        TransferResult<CFValue, UpperBoundStore> result = super.visitAssignment(node, in);
+
+        // When an existing array is assigned into, the store needs to be cleared -
```

Comment:
It only works in a subset of cases. I'll remove it, though - when I implemented it, I thought it would solve the problem, but we've since determined that it doesn't.